src/api/db.ts

```diff
@@ -1,47 +1,119 @@
 
 	return rows;
 }
 
+export async function getContent(id: number): Promise<Content|null> {
+	const [results] = await connection.execute(
+		"SELECT * FROM content WHERE content.id=? LIMIT 1",
+		[id]
+	) as [Content[], any];
+
+	return results[0];
+}
+
 export async function insertContent(content: Omit<Content, "id">): Promise<number> {
 	const [result] = await connection.execute(
-		"INSERT INTO content (type, name, content) VALUES (?, ?, ?);",
-		[content.type, content.name, content.content],
+		"INSERT INTO content (`type`, `name`, `content`) VALUES (?, ?, ?)",
+		[content.type, content.name, content.content]
 	) as [mysql.ResultSetHeader, any];
 
 	return result.insertId;
 }
 
+export async function updateContent(content: Content): Promise<void> {
+	await connection.execute(
+		"UPDATE content SET type=?, name=?, content=? WHERE id=?",
+		[content.type, content.name, content.content, content.id]
+	);
+}
+
+export async function getRelation(id: number): Promise<Relations|null> {
+	const [results] = await connection.execute(
+		"SELECT * FROM relations WHERE relations.id=? LIMIT 1",
+		[id]
+	) as [Relations[], any];
+
+	return results[0];
+}
+
 export async function insertRelation(relation: Omit<Relations, "id">): Promise<number> {
 	const [result] = await connection.execute(
-		"INSERT INTO relations (type, subtype, link_from, link_to, year_start, year_end, content) VALUES (?, ?, ?, ?, ?, ?, ?);",
-		[relation.type, relation.subtype, relation.link_from, relation.link_to, relation.year_start, relation.year_end || null, relation.content],
+		"INSERT INTO relations (type, subtype, link_from, link_to, year_start, year_end, content) VALUES (?, ?, ?, ?, ?, ?, ?)",
+		[relation.type, relation.subtype, relation.link_from, relation.link_to, relation.year_start, relation.year_end || null, relation.content]
 	) as [mysql.ResultSetHeader, any];
 
 	return result.insertId;
 }
 
+export async function updateRelation(relation: Relations): Promise<void> {
+	await connection.execute(
+		"UPDATE relations SET type=?, subtype=?, link_from=?, link_to=?, year_start=?, year_end=?, content=? WHERE id=?",
+		[relation.type, relation.subtype, relation.link_from, relation.link_to, relation.year_start, relation.year_end || null, relation.content, relation.id]
+	);
+}
+
+export async function getCitation(id: number): Promise<Citations|null> {
+	const [results] = await connection.execute(
+		"SELECT * FROM citations WHERE citations.id=? LIMIT 1",
+		[id]
+	) as [Citations[], any];
+
+	return results[0];
+}
 
 export async function insertCitation(citation: Omit<Citations, "id">): Promise<number> {
 	const [result] = await connection.execute(
-		"INSERT INTO citations (name, collection, content) VALUES (?, ?, ?);",
-		[citation.name, citation.collection, citation.content],
+		"INSERT INTO citations (name, collection, content) VALUES (?, ?, ?)",
+		[citation.name, citation.collection, citation.content]
```

Comment:
I thought we'd decided that it wasn't worth having a citations table, since no one's likely to want to query on a particular citation...?

Reply:
I remember discussing that, but I don't recall if that was a final decision or not (it doesn't seem to be in the planning document).

If we don't need to prevent duplicate citations or run queries on them directly, I could move them into the content/relation nodes, which would reduce some complexity.

Reply:
I haven't finished reviewing Ben's latest PR yet, but it looks from the commit messages like he was working on that alternate assumption. And yeah, this thing is already complicated enough! Let's talk about it in the morning.